### PR TITLE
[AArch64][GlobalISel] Support udot lowering for vecreduce add

### PR DESCRIPTION
**Pull request by chuongg3**
vecreduce_add(mul(ext, ext)) -> vecreduce_add(udot) 
vecreduce_add(ext) -> vecreduce_add(ext)

Vectors of scalar size of 8-bits with element count of multiples of 8

### CODE REVIEW

**Review by aemerson — COMMENTED** [review]
I didn't realize the predicates would work in the combiner tablegen files. Can you add a RUN line for GISel without +dotprod to check it doesn't produce udot in those cases? I suggest you precommit that test change and then rebase this patch on top of it.

File: llvm/lib/Target/AArch64/GISel/AArch64PreLegalizerCombiner.cpp [hunk 1]
```diff
@@ -228,6 +228,146 @@ void applyFoldGlobalOffset(MachineInstr &MI, MachineRegisterInfo &MRI,
       B.buildConstant(LLT::scalar(64), -static_cast<int64_t>(MinOffset)));
 }
 
+// Combines vecreduce_add(mul(ext, ext)) -> vecreduce_add(udot)
+// Or vecreduce_add(ext) -> vecreduce_add(ext)
+// Similar to performVecReduceAddCombine in SelectionDAG
+bool matchExtAddvToUdotAddv(MachineInstr &MI, MachineRegisterInfo &MRI) {
+  assert(MI.getOpcode() == TargetOpcode::G_VECREDUCE_ADD &&
```
Comment (davemgreen) [inline llvm/lib/Target/AArch64/GISel/AArch64PreLegalizerCombiner.cpp:235]: Can you add an assert that Subtarget->hadDotProd()

File: llvm/lib/Target/AArch64/GISel/AArch64PreLegalizerCombiner.cpp [hunk 2]
```diff
@@ -228,6 +228,146 @@ void applyFoldGlobalOffset(MachineInstr &MI, MachineRegisterInfo &MRI,
...
+  } else if (I1->getOpcode() == TargetOpcode::G_ZEXT ||
+             I1->getOpcode() == TargetOpcode::G_SEXT) {
+    Ext1SrcReg = I1->getOperand(1).getReg();
+    Ext2SrcReg = Builder.buildConstant(MRI.getType(Ext1SrcReg), 1)
+                     ->getOperand(0)
+                     .getReg();
+    DotOpcode = I1->getOpcode() == TargetOpcode::G_ZEXT ? AArch64::G_UDOT
+                                                        : AArch64::G_SDOT;
+  } else
+    return;
```
Comment (davemgreen) [inline llvm/lib/Target/AArch64/GISel/AArch64PreLegalizerCombiner.cpp:295]: Should this be impossible? llvm_unreachable?

File: llvm/lib/Target/AArch64/GISel/AArch64PreLegalizerCombiner.cpp [hunk 3]
```diff
@@ -228,6 +228,146 @@ void applyFoldGlobalOffset(MachineInstr &MI, MachineRegisterInfo &MRI,
...
+  if (NumOfVecReduce == 1) {
+    auto Zeroes = Builder.buildConstant(MidTy, 0)->getOperand(0).getReg();
+    auto Dot = Builder.buildInstr(DotOpcode, {MidTy},
+                                  {Zeroes, Ext1SrcReg, Ext2SrcReg});
+    Builder.buildVecReduceAdd(MI.getOperand(0), Dot->getOperand(0));
+  } else {
+    // Get the number of output vectors needed
+    SmallVector<LLT, 4> DotVecLLT;
+    auto SrcVecNum = SrcTy.getNumElements();
+    while (SrcVecNum - 16 >= 16 || SrcVecNum - 16 == 0) {
```
Comment (davemgreen) [inline llvm/lib/Target/AArch64/GISel/AArch64PreLegalizerCombiner.cpp:319]: Do you have tests for 24x vector types?

File: llvm/lib/Target/AArch64/GISel/AArch64PreLegalizerCombiner.cpp [hunk 4]
```diff
@@ -228,6 +228,146 @@ void applyFoldGlobalOffset(MachineInstr &MI, MachineRegisterInfo &MRI,
...
+    // Unmerge the source vectors
+    auto Ext1Unmerge = Builder.buildUnmerge(DotVecLLT, Ext1SrcReg);
+    auto Ext2Unmerge = Builder.buildUnmerge(DotVecLLT, Ext2SrcReg);
+
+    // Build the UDOT instructions
+    SmallVector<Register, 2> DotReg;
+    unsigned NumElements = 0;
+    for (unsigned i = 0; i < DotVecLLT.size(); i++) {
+      LLT ZeroesLLT;
+      // Check if it is 16 or 8 elements. Set Zeroes to the accoridng size
```
Comment (davemgreen) [inline llvm/lib/Target/AArch64/GISel/AArch64PreLegalizerCombiner.cpp:335]: according

File: llvm/lib/Target/AArch64/GISel/AArch64PreLegalizerCombiner.cpp [hunk 5]
```diff
@@ -228,6 +228,146 @@ void applyFoldGlobalOffset(MachineInstr &MI, MachineRegisterInfo &MRI,
...
+      DotReg.push_back(Builder
+                           .buildInstr(DotOpcode, {MRI.getType(Zeroes)},
+                                       {Zeroes, Ext1Unmerge.getReg(i),
+                                        Ext2Unmerge.getReg(i)})
+                           ->getOperand(0)
+                           .getReg());
+    }
+
+    // Merge the output
+    // auto a = MI.getOperand(1).getReg().changeNumElements(NumElements);
```
Comment (davemgreen) [inline llvm/lib/Target/AArch64/GISel/AArch64PreLegalizerCombiner.cpp:353]: Can be removed now

File: llvm/lib/Target/AArch64/GISel/AArch64PreLegalizerCombiner.cpp [hunk 6]
```diff
@@ -228,6 +228,146 @@ void applyFoldGlobalOffset(MachineInstr &MI, MachineRegisterInfo &MRI,
...
+        Builder.buildConcatVectors(LLT::fixed_vector(NumElements, 32), DotReg);
+
+    // Put it through a vector reduction
+    Builder.buildVecReduceAdd(MI.getOperand(0).getReg(),
+                              ConcatMI->getOperand(0).getReg());
+  }
+
+  // Erase the dead instructions
+  if (I1->getOpcode() == TargetOpcode::G_MUL) {
+    getDefIgnoringCopies(I1->getOperand(1).getReg(), MRI)->eraseFromParent();
```
Comment (davemgreen) [inline llvm/lib/Target/AArch64/GISel/AArch64PreLegalizerCombiner.cpp:364]: What happens if these have extra uses?

**Review by davemgreen — COMMENTED** [review]
The test look good to me, if you wanted to push those separately and didn't want to add some more. It might be worth making sure there are tests for 24x vector types and 48x vector types, if there are not already.

File: llvm/lib/Target/AArch64/GISel/AArch64PreLegalizerCombiner.cpp [hunk 7]
```diff
@@ -228,6 +228,146 @@ void applyFoldGlobalOffset(MachineInstr &MI, MachineRegisterInfo &MRI,
...
+        getDefIgnoringCopies(I1->getOperand(2).getReg(), MRI);
+    LLT Ext1DstTy = MRI.getType(ExtMI1->getOperand(0).getReg());
+    LLT Ext2DstTy = MRI.getType(ExtMI2->getOperand(0).getReg());
+
+    if (ExtMI1->getOpcode() != ExtMI2->getOpcode() || Ext1DstTy != Ext2DstTy)
+      return false;
+    I1Opc = ExtMI1->getOpcode();
+    SrcTy = MRI.getType(ExtMI1->getOperand(1).getReg());
+  } else
+    SrcTy = MRI.getType(I1->getOperand(1).getReg());
```
Comment (tschuett) [inline llvm/lib/Target/AArch64/GISel/AArch64PreLegalizerCombiner.cpp:261]: The braces are unbalanced.

File: llvm/lib/Target/AArch64/GISel/AArch64PreLegalizerCombiner.cpp [hunk 8]
```diff
@@ -228,6 +228,146 @@ void applyFoldGlobalOffset(MachineInstr &MI, MachineRegisterInfo &MRI,
...
+  LLT MidTy;
+  unsigned NumOfVecReduce;
+  if (SrcTy.getNumElements() % 16 == 0) {
+    NumOfVecReduce = SrcTy.getNumElements() / 16;
+    MidTy = LLT::fixed_vector(4, 32);
+  } else if (SrcTy.getNumElements() % 8 == 0) {
+    NumOfVecReduce = SrcTy.getNumElements() / 8;
+    MidTy = LLT::fixed_vector(2, 32);
+  } else
+    return;
```
Comment (tschuett) [inline llvm/lib/Target/AArch64/GISel/AArch64PreLegalizerCombiner.cpp:307]: The braces are unbalanced.

File: llvm/lib/Target/AArch64/GISel/AArch64PreLegalizerCombiner.cpp [hunk 9]
```diff
@@ -228,6 +228,199 @@ void applyFoldGlobalOffset(MachineInstr &MI, MachineRegisterInfo &MRI,
...
+      // Merge them all back to v16i8
+      NumOfVec = (NumOfVec + 1) / 2;
+      for (unsigned i = 0; i < NumOfVec; i++) {
+        Ext1UnmergeReg.push_back(
+            Builder
+                .buildMergeLikeInstr(
+                    LLT::fixed_vector(16, 8),
+                    {PadUnmergeDstReg1[i * 2], PadUnmergeDstReg1[(i * 2) + 1]})
+                ->getOperand(0)
+                .getReg());
```
Comment (davemgreen) [inline llvm/lib/Target/AArch64/GISel/AArch64PreLegalizerCombiner.cpp:368]: I think that ->getOperand(0).getReg() can be .getReg(0) from a MIBuilder. Here and 2 other places below.

**Review by davemgreen — APPROVED** [review]
Other than the comment below this LGTM, if no one else has any comments.

**Pull request merged**